File: scripts/daily-banner-update.ts
/**
 * Static Banner Configuration
 * 
 * This script sets up a static banner from Supabase storage.
 * Run once to initialize, or use for reference in your frontend code.
 */

import { createClient } from '@supabase/supabase-js';

// Initialize Supabase
const supabaseUrl = 'https://egzquylwclewcgpqnoig.supabase.co';
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || '';

if (!supabaseKey) {
  console.error('❌ SUPABASE_SERVICE_ROLE_KEY environment variable is not set');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

// Static banner configuration - Mobile optimized
const STATIC_BANNER = {
  imageUrl: 'https://egzquylwclewcgpqnoig.supabase.co/storage/v1/object/public/ef/Gemini_Generated_Image_3q1iy3q1iy3q1iy3.png',
  title: 'GTO Strategy Calculator',
  description: 'Optimize your OSM strategy with our advanced calculator tool',
  link: '/gto-calculator',
  isActive: true,
  // Responsive breakpoints for mobile optimization
  responsive: {
    mobile: { width: '100%', height: 'auto', maxHeight: '200px' },
    tablet: { width: '100%', height: 'auto', maxHeight: '280px' },
    desktop: { width: '100%', height: 'auto', maxHeight: '400px' }
  }
};

/**
 * Initialize static banner (run once)
 */
async function initializeStaticBanner(): Promise<boolean> {
  try {
    // Clear any existing banners
    await supabase
      .from('osm_banner')
      .delete()
      .neq('id', 0);

    // Insert static banner
    const { error } = await supabase
      .from('osm_banner')
      .insert([
        {
          image_url: STATIC_BANNER.imageUrl,
          title: STATIC_BANNER.title,
          description: STATIC_BANNER.description,
          link: STATIC_BANNER.link,
          is_active: STATIC_BANNER.isActive,
          is_static: true, // Flag to identify static banner
          responsive_config: STATIC_BANNER.responsive,
          created_at: new Date().toISOString(),
          updated_at: new Date().toISOString(),
        },
      ]);

    if (error) {
      console.error('❌ Supabase error:', error);
      return false;
    }

    console.log('✅ Static banner initialized successfully');
    console.log('📱 Mobile optimized: 100% width, auto height');
    return true;
  } catch (error) {
    console.error('❌ Error initializing banner:', error);
    return false;
  }
}

// Run once to set up
initializeStaticBanner().then(success => {
  process.exit(success ? 0 : 1);
});

File: scripts/daily_banner_update.py
#!/usr/bin/env python3
"""
Daily Banner Auto-Update Script for OSM Counter NG

This script runs daily to fetch new OSM scouting options and update the banner.
It uses an LLM to generate realistic young talented players for OSM.

Usage:
    python3 scripts/daily_banner_update.py

Schedule with cron:
    0 8 * * * cd /path/to/osm-counter-pwa && python3 scripts/daily_banner_update.py
"""

import json
import os
import sys
from datetime import datetime
from typing import Optional, Dict, Any

import requests
from openai import OpenAI


class BannerUpdater:
    """Handles daily banner updates for OSM Counter NG"""

    def __init__(self):
        """Initialize the banner updater with Supabase and OpenAI clients"""
        self.supabase_url = "https://egzquylwclewcgpqnoig.supabase.co"
        self.supabase_key = os.getenv("SUPABASE_SERVICE_ROLE_KEY")
        self.openai_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

        if not self.supabase_key:
            print("❌ SUPABASE_SERVICE_ROLE_KEY environment variable is not set")
            sys.exit(1)

    def generate_new_player(self) -> Optional[Dict[str, Any]]:
        """Generate a new OSM scouting player using LLM"""
        prompt = """You are an OSM (Online Soccer Manager) scouting expert. Generate a realistic young talented player that would be interesting for the OSM community.

Generate a JSON response with the following structure:
{
  "playerName": "Player's full name",
  "age": number between 15-22,
  "marketValue": "Market value in euros (e.g., €500K, €1.2M)",
  "position": "Position code (ST, CM, LB, CB, GK, etc.)",
  "rating": number between 6-9,
  "description": "Brief scouting report (1-2 sentences about potential and playing style)"
}

Make sure the player is realistic and interesting for OSM players. Vary the positions and nationalities."""

        try:
            print("🤖 Generating new player with LLM...")
            response = self.openai_client.chat.completions.create(
                model="gpt-4.1-mini",
                messages=[{"role": "user", "content": prompt}],
                temperature=0.7,
                max_tokens=300,
            )

            content = response.choices[0].message.content or ""

            # Extract JSON from the response
            start_idx = content.find("{")
            end_idx = content.rfind("}") + 1

            if start_idx == -1 or end_idx == 0:
                print("❌ Could not extract JSON from LLM response")
                return None

            json_str = content[start_idx:end_idx]
            player_data = json.loads(json_str)

            # Validate the data
            required_fields = [
                "playerName",
                "age",
                "marketValue",
                "position",
                "rating",
                "description",
            ]
            if not all(field in player_data for field in required_fields):
                print("❌ Invalid player data structure")
                return None

            # Generate a placeholder image URL
            player_data["imageUrl"] = (
                "https://i.ibb.co/tMSMxmwN/Gemini-Generated-Image-ticrt2ticrt2ticr.png"
            )

            return player_data

        except Exception as error:
            print(f"❌ Error generating player: {error}")
            return None

    def update_banner(self, player: Dict[str, Any]) -> bool:
        """Update the banner with a new player"""
        try:
            url = f"{self.supabase_url}/rest/v1/osm_banner"
            headers = {
                "apikey": self.supabase_key,
                "Authorization": f"Bearer {self.supabase_key}",
                "Content-Type": "application/json",
            }

            payload = {
                "player_name": player["playerName"],
                "age": player["age"],
                "market_value": player["marketValue"],
                "position": player["position"],
                "rating": player["rating"],
                "image_url": player["imageUrl"],
                "description": player["description"],
                "created_at": datetime.utcnow().isoformat(),
                "updated_at": datetime.utcnow().isoformat(),
            }

            response = requests.post(url, json=payload, headers=headers, timeout=10)

            if response.status_code not in [200, 201]:
                print(f"❌ Supabase error: {response.status_code} - {response.text}")
                return False

            print(
                f"✅ Banner updated with: {player['playerName']} ({player['age']} years old, {player['position']})"
            )
            return True

        except Exception as error:
            print(f"❌ Error updating banner: {error}")
            return False

    def run(self) -> None:
        """Main execution function"""
        print("🎯 Starting daily banner auto-update...")
        print(f"📅 Time: {datetime.utcnow().isoformat()}")

        # Generate new player
        new_player = self.generate_new_player()
        if not new_player:
            print("❌ Failed to generate new player")
            sys.exit(1)

        print(f"📊 Generated player: {new_player['playerName']}")
        print(f"   Age: {new_player['age']}")
        print(f"   Position: {new_player['position']}")
        print(f"   Rating: {new_player['rating']}/10")
        print(f"   Market Value: {new_player['marketValue']}")

        # Update banner
        print("💾 Updating banner in Supabase...")
        success = self.update_banner(new_player)

        if success:
            print("✅ Daily banner update completed successfully!")
            sys.exit(0)
        else:
            print("❌ Failed to update banner")
            sys.exit(1)


if __name__ == "__main__":
    updater = BannerUpdater()
    updater.run()
